contributors/design-proposals/customresources-validation.md

```diff
@@ -0,0 +1,525 @@
+                                "baz"
+                            ],
+                            "additionalProperties": false
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+### Error messages
+
+The following examples illustrate the type of validation errors generated by using the go-openapi validate library.
+
+The description is not taken into account, but a better error output can be easily [added](https://github.com/go-openapi/errors/blob/master/headers.go#L23) to go-openapi.
+
+1. `data.foo in body should be at least 4 chars long`
+2. `data.foo in body should be greater than or equal to 10`
+3. `data.foo in body should be one of [bar baz]`
+4. `data.foo in body must be of type integer: "string"`
+5. `data.foo in body should match '^[a-zA-Z0-9_]*$'`
+6. `data.foo in body is required`
+7. When foo validates if it is a multiple of 3 and 5:
+```
+data.foo in body should be a multiple of 5
+data.foo in body should be a multiple of 3
+must validate all the schemas (allOf)
+```
+
+## Validation Behavior
+
+The schema will be described in the `CustomResourceDefinitionSpec`. The validation will be carried out using the [go-openapi validation library](https://github.com/go-openapi/validate).
+ 
+While creating/updating the CR, the metadata is first validated. To validate the CR against the spec in the CRD, we _must_ have server-side validation and we _can_ have client-side validation.
+
+### Metadata
+
+ObjectMeta and TypeMeta are implicitly specified. They do not have to be added to the JSON-Schema of a CRD. The validation already happens today as part of the apiextensions-apiserver REST handlers.
+
+### Server-Side Validation
+
+The server-side validation is carried out after sending the request to the apiextensions-apiserver, i.e. inside the CREATE and UPDATE handlers for CRs.
+
+We do a schema pass there using the https://github.com/go-openapi/validate validator with the provided schema in the corresponding CRD. Validation errors are returned to the caller as for native resources.
+
+JSON-Schema also allows us to reject additional fields that are not defined in the schema and only allow the fields that are specified. This can be achieved by using `"additionalProperties": false` in the schema. However, there is danger in allowing CRD authors to set `"additionalProperties": false` because it breaks version skew (new client can send new optional fields to the old server). So we should not allow CRD authors to set `"additionalProperties": false`.
+
+### Client-Side Validation
+
+The client-side validation is carried out before sending the request to the api-server, or even completely offline. This can be achieved while creating resources through the client i.e. kubectl using the --validate option.
+
+If the API type serves the JSON-Schema in the swagger spec, the existing kubectl code will already be able to also validate CRs. This will be achieved as a follow-up.
```

Comment:
Sorry that I see this late. I have two points here:

1. If we register CRDs with aggregation server (which I think we should), this will be aggregated automatically.
2. OpenAPI supports a subset of json schema. I think instead of asking users to provide JSON Schema we should ask them to provide OpenAPI spec that has a Definition model in it (that is a subset of JSON Schema).